### PR TITLE
Add draft RFC for default git branch protection rules

### PR DESCRIPTION
**Pull request by AP-Hunt**
Following from discussion at the TOC meeting on 4th October 2022, I've written up a draft RFC to formalise the discussion around branch protection rules which has happened in the TOC meeting and in #438.

I've added @Benjamintf1 and @rkoster as authors to acknowledge that while I put the words together, the ideas are theirs.

### CODE REVIEW

**Comment by beyhan** [comment]
My personal opinion is:
* PR only workflow for approvers (If we could find a way to simplify this when there are CI issues would be great)
* More simplicity for bots

**Comment by rkoster** [comment]
```
Every change in the revision’s history was agreed to by two trusted persons prior to submission, and both of these trusted persons were strongly authenticated. (Exceptions from [Verified History](https://slsa.dev/spec/v0.1/requirements#verified-history) apply here as well.)

The following combinations are acceptable:
Uploader and reviewer are two different trusted persons.
Two different reviewers are trusted persons.
```
https://slsa.dev/spec/v0.1/requirements#two-person-reviewed

According to SLSA it seems like PRs submitted by an Approver would only need 1 additional approval. However, I don't think this is something that could be enforced with GitHub branch protection rules. So better to just require 2 approvals for everybody.

**Comment by ameowlia** [comment]
I agree with @Benjamintf1. 2 approvers sounds like _A LOT_, especially given that 95% of our PRs are written _by_ working group approvers, but when you open a PR you cannot approve your own PR. This means 3 people would be required for each PR (1 writer + 2 approvers).

In ARP we do not have any problems that could be solved by multiple approvers. Before we require multiple approvers I would want to hear a compelling reason why it is necessary.

### Example 1
* ARP Diego has 8 approvers
* Two are currently on parental leave
* If two approvers are required per PR...
  * Approver-1 writes the PR
  * Approver-2 approves the PR
  * Approver-3 approvers and merges the PR
* ⚠️ This means 50% of our current approvers who aren't on leave will be involved in every change that the team wants to make. 


### Example 2
* ARP Docs has 2 approvers
* If one approver is required per PR...
   * Approver-1 writes the PR
   * Approver-2 approves the PR
* This means _every_ PR this team makes will require participation from both approvers.
* ⚠️ This means when one member is on vacation, the other cannot get work in (without reaching out to me as the admin).

### Admin as workround is not sustainable
As the only admin for the ARP repos I have already been pinged 20+ times in the past week to fix things on an emergency basis. I worry that if approvers are required then when there is a time crunch I will be paged more often to use my admin powers.

### My request
**Make the number of approvers required configurable.** Small teams (like the docs teams above) shouldn't be forced to have _any_ approvers. And I see no reason why other teams should require more than one approver.

**Comment by Gerg** [comment]
I oppose requiring more than one PR review. IMO, [diffusion of responsibility](https://en.wikipedia.org/wiki/Diffusion_of_responsibility) reduces the quality of multi-review PRs.

**Comment by jrussett** [comment]
I don't think that adding another reviewer will improve the quality of the PR and will instead just increase the time it takes to get work done. It feels like we struggle enough to make sure that at least one person reviews the PR after another person has already done the acceptance, why add yet another layer of review and spent time?

**Comment by AP-Hunt** [comment]
I've updated the RFC in response to feedback and discussion among the TOC.

Changes:
* Only bots will be allowed to push directly to the branch
* Humans must use a PR workflow
* WG with <=3 approvers will require 0 approvals, WG with >3 will require 1 approval.

I'll squash the commits before it gets merged.

**Review by rkoster — APPROVED** [review]

**Comment by AP-Hunt** [comment]
@cloudfoundry/wg-leads Final comment period on this PR starts today, ends 15th November.

**Review by beyhan — APPROVED** [review]

**Comment by pivotal-marcela-campo** [comment]
I am not comfortable with the problems it may cause to enforce this. The working group definition doesn’t have a calendar attached and we are not removing approvers each time someone takes time off or pauses their work on open source projects. 

Example:
Cloud Service Brokers team in Service management WG currently has 5 approvers

By the end of this month, one is going on leave, another one on holidays and we have a member on rotation on another team. This means we are left with 2 active approvers temporarily and hence wouldn’t require approvals, but we do because the WG definition doesn’t have a calendar attached. If one of them gets ill, or decides to take a day off then no work can be possibly merged. 

I do see the recommendations valuable and definitely think we should all follow whenever possible.

**Review by stephanme — APPROVED** [review]
+1 for the concept.

A minor changed needed for the version branches (remove one v).

Regarding the implementation: the branchprotector tool needs some enhancements before we can use it. At the moment it doesn't allow to configure teams that can bypass the PR flow. This is needed for bots.

File: toc/rfc/rfc-draft-branch-protection.md
Comment (emalm) [inline toc/rfc/rfc-draft-branch-protection.md:12]: Fixing some typos:
```suggestion
In Q3 2022, the Technical Oversight Committee accepted and implemented [an RFC for the management of GitHub teams and user access](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0014-github-teams-and-access.md). This unintentionally broke the workflow and branch protection rules of some working groups.
```

File: toc/rfc/rfc-draft-branch-protection.md
Comment (emalm) [inline toc/rfc/rfc-draft-branch-protection.md:14]:
```suggestion
This RFC seeks to address that by defining a baseline of branch protection rules across all repositories under management in the cloudfoundry GitHub org.
```

File: toc/rfc/rfc-draft-branch-protection.md
Comment (emalm) [inline toc/rfc/rfc-draft-branch-protection.md:29]:
```suggestion
We propose that protection be applied to all branches matching either the default branch or the expression `v[0-9]*`.
```

**Review by emalm — CHANGES_REQUESTED** [review]
In addition to these editorial corrections, let's discuss whether we should amend the RFC to include @stephanme's [idea](https://github.com/cloudfoundry/community/pull/449#issuecomment-1315399746) of allowing per-repository override of these default branch-protection rules.

File: toc/rfc/rfc-draft-branch-protection.md
Comment (emalm) [inline toc/rfc/rfc-draft-branch-protection.md:32]:
```suggestion
* only the bots for the Working Group owning the repository and any areas in the Working Group contributing to the repository may push directly to the branch
```

**Review by emalm — APPROVED** [review]

**Pull request merged**
